src/services/domain/roles/roles.service.ts

```diff
@@ -303,65 +395,64 @@ export class RolesService {
-          state,
-          application.id,
-          community.hubID, // Store the hub the application is for, regardless of level
-          application.createdDate,
-          application.updatedDate
-        );
-
-        const communityParent = await this.communityService.getParentCommunity(
-          community
-        );
-
-        // not an hub
-        if (communityParent) {
-          // For Challenge or an Opportunity, need to dig deeper...
-          const challengeForCommunity = await this.challengeRepository
-            .createQueryBuilder('challenge')
-            .leftJoinAndSelect('challenge.community', 'community')
-            .orWhere('community.id like :communityID')
-            .setParameters({ communityID: `%${community.id}%` })
-            .getOne();
-          if (challengeForCommunity) {
-            // the application is issued for a challenge
-            applicationResult.challengeID = challengeForCommunity.id;
-          } else {
-            const opportunityForCommunity = await this.opportunityRepository
-              .createQueryBuilder('opportunity')
-              .leftJoinAndSelect('opportunity.challenge', 'challenge')
-              .leftJoinAndSelect('opportunity.community', 'community')
-              .orWhere('community.id like :communityID')
-              .setParameters({ communityID: `%${community.id}%` })
-              .getOne();
-
-            if (
-              !opportunityForCommunity ||
-              !opportunityForCommunity.challenge
-            ) {
-              throw new RelationshipNotFoundException(
-                `Unable to find Challenge or Opportunity with the community specified: ${community.id}`,
-                LogContext.COMMUNITY
-              );
-            }
-
-            // the application is issued for an an opportunity
-            applicationResult.opportunityID = opportunityForCommunity.id;
-            applicationResult.challengeID =
-              opportunityForCommunity.challenge.id;
-          }
-        }
+        const applicationResult =
+          await this.buildApplicationResultForCommunityApplication(
+            community,
+            state,
+            application
+          );
 
         applicationResults.push(applicationResult);
       }
     }
     return applicationResults;
   }
 
-  private generateRandomSuffix(): string {
-    return Math.floor(Math.random() * 100000).toString();
+  private async buildApplicationResultForCommunityApplication(
+    community: ICommunity,
+    state: string,
+    application: IApplication
+  ): Promise<ApplicationForRoleResult> {
+    const applicationResult = new ApplicationForRoleResult(
+      community.id,
+      community.displayName,
+      state,
+      application.id,
+      community.hubID,
+      application.createdDate,
+      application.updatedDate
+    );
+
+    const isHubCommunity = await this.communityService.isHubCommunity(
+      community
+    );
+
+    if (!isHubCommunity) {
```

Comment:
I would suggest using if-guard here and in the nested ifs to improve readability.